lib/ash/query/function/ago.ex

```diff
@@ -3,28 +3,66 @@ defmodule Ash.Query.Function.Ago do
   Subtracts the given interval from the current time in UTC.
 
   For example:
-     deleted_at > ago(7, :days)
+     deleted_at > ago(7, :day)
 
   Documentation + available intervals inspired by the corresponding ecto interval implementation
   """
+
   use Ash.Query.Function, name: :ago
 
   def args, do: [[:integer, :duration_name]]
 
   def evaluate(%{arguments: [factor, interval]}) do
-    interval =
-      case interval do
-        :year -> :years
-        :month -> :months
-        :week -> :weeks
-        :day -> :days
-        :hour -> :hours
-        :minute -> :minutes
-        :second -> :seconds
-        :millisecond -> :milliseconds
-        :microsecond -> :microseconds
+    now = DateTime.utc_now()
+    shifted = datetime_add(now, -factor, interval)
+    {:known, shifted}
+  end
+
+  @doc false
+  def datetime_add(datetime, amount_to_add, unit)
+      when unit in [:microsecond, :millisecond, :second] do
+    DateTime.add(datetime, amount_to_add, unit)
+  end
+
+  def datetime_add(datetime, amount_to_add, :minute) do
+    DateTime.add(datetime, amount_to_add * 60, :second)
+  end
+
+  def datetime_add(datetime, amount_to_add, :hour) do
+    DateTime.add(datetime, amount_to_add * 60 * 60, :second)
+  end
+
+  def datetime_add(datetime, amount_to_add, :day) do
+    date = Date.add(datetime, amount_to_add)
+    %{datetime | year: date.year, month: date.month, day: date.day}
+  end
+
+  def datetime_add(datetime, amount_to_add, :week) do
+    datetime_add(datetime, 7 * amount_to_add, :day)
+  end
+
+  def datetime_add(datetime, amount_to_add, :month) do
+    months_since_zero = datetime.year * 12 + datetime.month - 1 + amount_to_add
+    year = div(months_since_zero, 12)
+    {year, month} =
+      case rem(months_since_zero, 12) do
+        x when x < 0 -> {year - 1, 12 + x + 1}
+        x -> {year, x + 1}
       end
 
-    {:known, Timex.shift(Timex.now(), [{interval, factor * -1}])}
+    # clamp to last day of month
+    last_day = Calendar.ISO.days_in_month(year, month)
+    day = min(datetime.day, last_day)
+    %{datetime | year: year, month: month, day: day}
+  end
+
+  def datetime_add(datetime = %{month: month, day: day}, amount_to_add, :year) do
+    result = %{datetime | year: datetime.year + amount_to_add}
+    # Adjust 29-Feb to 01-March when not a leap year
+    if month == 2 and day == 29 and not Calendar.ISO.leap_year?(result.year) do
```

Comment:
Preserving the Timex behaviour here. Shifting 29th Feb by years to a non-leapyear adjusts forward to 1st of March.